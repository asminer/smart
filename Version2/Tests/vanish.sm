
// $Id$

/*
    Test of PNs with vanishing markings
*/


pn foo(int n) := {
  place p1, p2, p3, p4, p5, p6, p7, p8;
  init(p1:n);

  trans T1, T2, T3, T4, t5, t6, T7, t8, t9, T10, t11;

  firing(
	T1 : expo(1.0),
	T2 : expo(2.1),
	T3 : expo(3.2),
	T4 : expo(4.0),
	t5 : 0,
	t6 : 0,
	T7 : expo(7.1),
	t8 : 0,
	t9 : 0,
	T10 : expo(10.0),
	t11 : 0
  );

  weight(t5 : 1.0, t6 : 1.0);
  weight(t8 : 1.0, t9 : 1.0);
  // t11 by itself, shouldn't need weight

  arcs( 
	p2 : T1, T1 : p1,
	p1 : T2, T2 : p2,
	p1 : T3, T3 : p3,
	p4 : T4, T4 : p2,
	p3 : t5, t5 : p4,
	p3 : t6, t6 : p5,
	p6 : T7, T7 : p4,
	p5 : t8, t8 : p6,
	p5 : t9, t9 : p7,
	p8 : T10, T10 : p6,
	p7 : t11, t11 : p8
  );

  int ns := num_states(true);
  int mc := num_arcs(true);
};

#PnMarkingStyle SAFE

print("Tangible only:\n", foo(1).ns, "\nMarkov chain:\n", foo(1).mc, "\n");

print("Resetting, ignore this:", foo(0).ns, "\n");

print("Building MC & RS simultaneously, should get same Markov chain:\n",
      foo(1).mc, "\n");


//
// Test of vanishing loops
//

pn bar(bool can_escape) := {
  place p1, p2, p3, p4, p5, p6;
  init(p1:1);

  trans T1, t2, t3, t4, t5, Tt6, T7;
  
  firing(
	T1 : expo(2.0),
  	t2 : 0,
	t3 : 0,
	t4 : 0,
	t5 : 0,
	T7 : expo(7.1)
  );
  cond(can_escape,
	firing(Tt6 : 0); weight(t5 : 1.0, Tt6 : 1.0),
 	firing(Tt6 : expo(6.1))
  );
  arcs(
	p1 : T1, T1 : p3,
	p3 : t2, t2 : p5,
	p5 : t3, t3 : p6,
	p6 : t4, t4 : p4,
	p4 : t5, t5 : p3,
	p4 : Tt6, Tt6 : p2,
	p2 : T7, T7 : p1
  );
  
  int ns := num_states(true);
  int mc := num_arcs(true);
};

print("A Vanishing Loop that can be escaped\n");
print("Tangible:\n", bar(true).ns, "\nMarkov chain:\n", bar(true).mc, "\n");

print("A Vanishing Loop that cannot be escaped\n");
print("Tangible:\n", bar(false).ns, "\nMarkov chain:\n", bar(false).mc, "\n");


print("Same test for MC & RS simultaneously\n");

print("A Vanishing Loop that can be escaped\n");
print("Markov chain:\n", bar(true).mc, "\n");
print("A Vanishing Loop that cannot be escaped\n");
print("Markov chain:\n", bar(false).mc, "\n");
