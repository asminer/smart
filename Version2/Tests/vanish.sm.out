Tangible only:
State 0: [p1]
State 1: [p2]
State 2: [p4]
State 3: [p6]
State 4: [p8]
5
Markov chain:
Markov chain:
Column 0:
	1 : 1
Column 1:
	0 : 2.1
	2 : 4
Column 2:
	0 : 1.6
	3 : 7.1
Column 3:
	0 : 0.8
	4 : 10
Column 4:
	0 : 0.8
8
Resetting, ignore this:WARNING:
	Assuming zero initial marking in PN foo
State 0: []
1
Building MC & RS simultaneously, should get same Markov chain:
Markov chain:
Column 0:
	1 : 1
Column 1:
	0 : 2.1
	2 : 4
Column 2:
	0 : 1.6
	3 : 7.1
Column 3:
	0 : 0.8
	4 : 10
Column 4:
	0 : 0.8
8
A Vanishing Loop that can be escaped
Tangible:
State 0: [p1]
State 1: [p2]
2
Markov chain:
Markov chain:
Column 0:
	1 : 7.1
Column 1:
	0 : 1.99997
2
A Vanishing Loop that cannot be escaped
Tangible:
State 0: [p1]
1
Markov chain:
ERROR in model bar instantiated in file vanish.sm near line 102:
	Inescapable loop of vanishing states
null
Same test for MC & RS simultaneously
A Vanishing Loop that can be escaped
Markov chain:
Markov chain:
Column 0:
	1 : 7.1
Column 1:
	0 : 1.99997
2
A Vanishing Loop that cannot be escaped
Markov chain:
ERROR in model bar instantiated in file vanish.sm near line 110:
	Inescapable loop of vanishing states
null
